app/api/events.py

```diff
@@ -586,7 +619,229 @@ async def confirmation(request: Request, id: str, token: AccessTokenPayload = De
+    """ Give all members absent on binding event a penalty """
+    db = get_database(request)
+    event = get_event_or_404(db, id)
+
+    if not event["bindingRegistration"]:
+        raise HTTPException(400, "Cannot penalize on non-binding events")
+    
+    if not event_has_started(event):
+        raise HTTPException(400, "Cannot penalize before event start") 
+    
+    if not event["confirmed"]:
+        raise HTTPException(400, "Cannot penalize unconfirmed event")
+        
+
+    # Aggregate all participants that are confirmed, but not attended,
+    # and group them by ids
+    pipeline = [
+        {"$match": {"eid": event["eid"]}},
+        {"$unwind": {"path": "$participants"}},
+        {"$match": {"participants.confirmed": {"$eq": True}}},
+        {"$match": {"participants.attended": {"$ne": True}}},
+        {"$group": {"_id": "$participants.id"}},
+    ]
+    absent = list(db.events.aggregate(pipeline))
+
+    if not absent or len(absent) == 0:
+        raise HTTPException(400, "No members to penalize")
+    
+    # Get ids
+    absent_ids = [p["_id"] for p in absent]
+    
+    # Get already penalized ids
+    penalized_ids = event["registeredPenalties"]
+
+    # Members to penalize
+    to_penalize = []
+    for m_id in absent_ids:
+        if m_id not in penalized_ids:
+            to_penalize.append(m_id)
+
+
+    if len(to_penalize) == 0:
+        raise HTTPException(400, "Members already penalized")
+
+    # Add penalty to selected
+    updates = []
+    for m_id in to_penalize:
+        # Add penalty to member and all event participant entries
+        await penalize(db, m_id)
+        # Add id to registeredPenalties on event
+        updates.append(UpdateOne(
+            {"eid": event["eid"]}, 
+            {"$addToSet": {"registeredPenalties": m_id}}
+        ))
 
+    # Write to event db
+    res = db.events.bulk_write(updates)
+
+    if not res:
+        raise HTTPException(500)
+
+
+
+@router.post('/{id}/qr', dependencies=[Depends(validate_uuid)])
+def create_registration_qr(request: Request, id: str, token: AccessTokenPayload = Depends(authorize_admin)):
+    """ Generate a registration QR code for event """
+    db = get_database(request)
+    event = get_event_or_404(db, id)
+
+    # Check whether QR already has been created
+    rid_created = 'register_id' in event and event['register_id']
+    filepath = f'{get_qr_path(request)}/{event["eid"].hex}.pdf'
+    if rid_created and os.path.exists(filepath):
+        raise HTTPException(400, "Event QR already created")
+    
+
+    # Generate new unique id for attendence registration
+    # Only if rid has not yet been created
+    register_id = uuid4()
+
+    res = db.events.update_one(
+        {'eid': UUID(id)},
+        {'$set': {'register_id': register_id}}
+    )
+
+    if not res:
+        raise HTTPException(500)
+    
+    # Create directory if not present
+    if not os.path.exists(get_qr_path(request)):
```

Comment:
Can we save the return value of `get_qr_path` and reuse it in makedirs and further down?